cephfs/statx.go

```diff
@@ -0,0 +1,133 @@
+package cephfs
+
+/*
+#cgo LDFLAGS: -lcephfs
+#cgo CPPFLAGS: -D_FILE_OFFSET_BITS=64
+#include <cephfs/libcephfs.h>
+*/
+import "C"
+
+// StatxWant values contain bit-flags indicating what data should be
+// populated by a statx-type call.
+type StatxWant uint32
+
+const (
+	// StatxMode requests the mode value be filled in.
+	StatxMode = StatxWant(C.CEPH_STATX_MODE)
+	// StatxNlink requests the nlink value be filled in.
+	StatxNlink = StatxWant(C.CEPH_STATX_NLINK)
+	// StatxUid requests the uid value be filled in.
+	StatxUid = StatxWant(C.CEPH_STATX_UID)
+	// StatxRdev requests the rdev value be filled in.
+	StatxRdev = StatxWant(C.CEPH_STATX_RDEV)
+	// StatxAtime requests the access-time value be filled in.
+	StatxAtime = StatxWant(C.CEPH_STATX_ATIME)
+	// StatxMtime requests the modified-time value be filled in.
+	StatxMtime = StatxWant(C.CEPH_STATX_MTIME)
+	// StatxIno requests the inode be filled in.
+	StatxIno = StatxWant(C.CEPH_STATX_INO)
+	// StatxSize requests the size value be filled in.
+	StatxSize = StatxWant(C.CEPH_STATX_SIZE)
+	// StatxBlocks requests the blocks value be filled in.
+	StatxBlocks = StatxWant(C.CEPH_STATX_BLOCKS)
+	// StatxBasicStats requests all the fields that are part of a
+	// traditional stat call.
+	StatxBasicStats = StatxWant(C.CEPH_STATX_BASIC_STATS)
+	// StatxBtime requests the birth-time value be filled in.
+	StatxBtime = StatxWant(C.CEPH_STATX_BTIME)
+	// StatxVersion requests the version value be filled in.
+	StatxVersion = StatxWant(C.CEPH_STATX_VERSION)
+	// StatxAllStats requests all known stat values be filled in.
+	StatxAllStats = StatxWant(C.CEPH_STATX_ALL_STATS)
+)
+
+// AtFlags represent flags to be passed to calls that control how files
+// are used or referenced. For example, not following symlinks.
+type AtFlags uint
+
+const (
+	// AtNoAttrSync requests that the stat call only fetch locally-cached
+	// values if possible, avoiding round trips to a back-end server.
+	AtNoAttrSync = AtFlags(C.AT_NO_ATTR_SYNC)
+	// AtSymlinkNofollow indicates the call should not follow symlinks
+	// but operate on the symlink itself.
+	AtSymlinkNofollow = AtFlags(C.AT_SYMLINK_NOFOLLOW)
+)
+
+// CephStatx instances are returned by extended stat (statx) calls.
+// Note that CephStatx results are similar to but not identical
+// to (Linux) system statx results.
+type CephStatx struct {
+	statx C.struct_ceph_statx
+}
+
+// Mask returns a value containing the bits indicating what
+// statx fields have been set.
+func (csx *CephStatx) Mask() StatxWant {
```

Comment:
StatxWant is a weird type name in this case. The user may have requested attributes that could not be fetched, so `stx_mask` would not have all the 'want' bits, but only the ones that actually got returned.

Maybe add some user-friendly glue similar to `rbd/features.go`?

Reply:
I should probably just go with StatxMask to better match the C naming. It's not like bitmasks are unheard of in Go.
Interestingly, the call is allowed to return fields & mask bits the user did not ask for. But you are also right about the negative case.
Some of that glue could be good here but I think the 99% use case is passing StatxBasicStats and getting back at least those bits set.
If there's no objection I'll rename to StatxMask and look into adding more glue in a later PR.

Reply:
+1 on the StatxWant -> StatxMask

Just calling `Statx()` with `StatxBasicStats` does not make much sense, in that case `Stat()` can be called instead and is shorter/easier. I would expect users of `Statx()` to always use additional flags.

Reply:
Hehe. I should have pointed this out earlier, but forgot. In the cephfs api there is no stat! 

The api only has statx. I am assuming this is the case because statx is a superset of stat.
I'd be open to go-ceph implementing a "simpler" stat wrapper that uses statx underneath but I wanted to start with the "thinner" layer as we often try to do in this project.

Reply:
hmm, really? I see at least `ceph_stat`, `ceph_fstat` and `ceph_lstat` in cephfs.h.

There is also `ceph_setattrx` that indicates that setting `CephStatx` members should be needed. With that, I am getting more convinces that exposing members of `CephStatx` makes sense (and copying them from/to `ceph_statx`).

Reply:
Odd. I'm looking at src/include/cephfs/libcephfs.h .... 
Hahah. These are new in octopus! I use a locally checked out ceph of the latest nautilus and did not see such a thing.

Looks like the change happened here: https://github.com/ceph/ceph/pull/25650
and skimming that thread I feel like I just stepped on a landmine. I'm going to study it a bit closer but I'm not feeling good about this overall now.

Reply:
Oh, that is interesting! Maybe leave the basic `ceph_stat()` out after all, at least for now. (FWIW, I'm on Fedora 31 with ceph-14.2.8.)

Just keep in mind that `ceph_setattrx()` will be a useful addition, and that uses a `CephStatx` with fields set by the user.

Reply:
Thanks! The latter is a good point which I had overlooked until you pointed it out.